pennylane/capture/transforms/capture_decompose.py

```diff
@@ -0,0 +1,126 @@
+
+        if isinstance(gate_set, (str, type)):
+            gate_set = set([gate_set])
+
+        if isinstance(gate_set, Iterable):
+            gate_types = tuple(gate for gate in gate_set if isinstance(gate, type))
+            gate_names = set(gate for gate in gate_set if isinstance(gate, str))
+            self.gate_set = lambda op: (op.name in gate_names) or isinstance(op, gate_types)
+        else:
+            self.gate_set = gate_set
+
+        super().__init__()
+
+    def stopping_condition(self, op: qml.operation.Operator) -> bool:
+        """Function to determine whether or not an operator needs to be decomposed or not.
+
+        Args:
+            op (qml.operation.Operator): Operator to check.
+
+        Returns:
+            bool: Whether or not ``op`` is valid or needs to be decomposed. ``True`` means
+            that the operator does not need to be decomposed.
+        """
+        if not op.has_decomposition:
+            if not self.gate_set(op):
+                warnings.warn(
+                    f"Operator {op.name} does not define a decomposition and was not "
+                    f"found in the target gate set. To remove this warning, add the operator name "
+                    f"({op.name}) or type ({type(op)}) to the gate set.",
+                    UserWarning,
+                )
+            return True
+        return self.gate_set(op)
+
+    def decompose_operation(self, op: qml.operation.Operator):
+        """Decompose a PennyLane operation instance if it does not satisfy the
+        provided gate set.
+
+        Args:
+            op (Operator): a pennylane operator instance
+
+        Returns:
+            Any
+
+        This method is only called when the operator's output is a dropped variable,
+        so the output will not affect later equations in the circuit.
+
+        See also: :meth:`~.interpret_operation_eqn`, :meth:`~.interpret_operation`.
+        """
+        if self.gate_set(op):
+            return self.interpret_operation(op)
+
+        qml.capture.disable()
+        try:
+            decomposition = list(
+                _operator_decomposition_gen(
+                    op, self.stopping_condition, max_expansion=self.max_expansion
+                )
+            )
+         finally:
+            qml.capture.enable()
+
+        return [self.interpret_operation(decomp_op) for decomp_op in decomposition]
+
+    def interpret_operation_eqn(self, eqn):
+        """Interpret an equation corresponding to an operator.
+
+        Args:
+            eqn (jax.core.JaxprEqn): a jax equation for an operator.
+
+        See also: :meth:`~.interpret_operation`.
+
+        """
+        invals = (self.read(invar) for invar in eqn.invars)
+        with qml.QueuingManager.stop_recording():
+            op = eqn.primitive.impl(*invals, **eqn.params)
+        if eqn.outvars[0].__class__.__name__ == "DropVar":
+            return self.decompose_operation(op)
+        return op
```

Comment:
If the transformed function returns an op that needs to be decomposed and we return all the ops from the decomposition, how are we supposed to handle that? How will the calling function use this?

Reply:
So something like:

```
def f():
    qml.IsingXX(0.5, wires=(0,1))
    return qml.Rot(1.2, 2.3, 3.4, wires=0)

def g():
    op = f()
    qml.adjoint(op)
```

Ugk. That is one annoying edge case.

Reply:
So maybe we do this now, but figure out a better way to handle the above edge case next quarter?

Reply:
I personally don't see this to be a major issue.

Case 1: a function that is called inside the function being transformed returns an op, (if we transform `g` in the above example). In this case, we don't transform `Rot` when it is returned, but because `qml.adjoint` is a `DropVar`, we _do_ transform `qml.adjoint(Rot)`, which is what the final behaviour should've been in the first place.

Case 2: a function being transformed returns an op (if we transform `f` in your above example). I'd argue this is a meaningless "workflow". Program capture only makes sense if the full workflow is being captured, so ops that are returned by a function are only meaningful if they are consumed/dropped in a workflow that is also being captured/interpreted.